### ISSUE TITLE
github oauth no longer works after adding "no-api" to private repository.

### ISSUE THREAD

**Issue by donquixote** [composer/composer#9679]
## TLDR
I work on a project that gets one of its packages from a private github repository.
Recently I added the "no-api" setting to the repository.

Without the "no-api", the authentication was using a github oauth token.
After adding the "no-api" setting, it asks me for a github username, instead of using the oauth token.

Why might this be?

## Details following the issue template
My `composer.json`:

**v1**
```json
{
    [..]
    "repositories": {
        "<repo name>": {
            "type": "vcs",
            "url": "https://github.com/<vendor>/<private repo>"
        }
    },
```
**v2**
```json
{
    [..]
    "repositories": {
        "<repo name>": {
            "type": "vcs",
            "no-api": true,
            "url": "https://github.com/<vendor>/<private repo>"
        }
    },
```

Output of `composer diagnose`:

```
<some notices about unrelated packages>
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Composer version: 1.10.19
PHP version: 7.3.24
PHP binary path: /usr/bin/php7.3
OpenSSL version: OpenSSL 1.1.1g  21 Apr 2020
```

When I run this command: `composer update --lock -vvv`:

I get the following output in the two cases:
(I censor this for privacy / confidentiality reasons)

**v1**
(without "no-api")
Composer uses the oauth token (or asks me for a new one)
```
[...]
Loading config file /var/www/html/auth.json
[...]
Loading composer repositories with package information
[..]
Using GitHub token authentication
Downloading https://api.github.com/repos/<vendor>/<private repo>
[..]
```
**v2**
(with "no-api")
Composer asks for github username (+ password, I suppose)
```
[...]
Loading config file /var/www/html/auth.json
[...]
Loading composer repositories with package information
[...]
Executing command (/home/<user>/.composer/cache/vcs/https---github.com-<vendor>-<repo>/): git rev-parse --git-dir
Executing command (/home/<user>/.composer/cache/vcs/https---github.com-<vendor>-<repo>/): git remote -v
Executing command (/home/<user>/.composer/cache/vcs/https---github.com-<vendor>-<repo>/): git remote set-url origin 'https://github.com/<vendor>/<repo>' && git remote update --prune origin && git remote set-url origin 'https://github.com/<vendor>/<repo>'
Username for 'https://github.com': 
```

**Expected**
I would expect the oauth token to work in both cases.

**Comment by donquixote** [comment]
Interesting..
It seems that this works:

```json
{
    [..]
    "repositories": {
        "<repo name>": {
            "type": "vcs",
            "no-api": true,
            "url": "git@github.com:<vendor>/<private repo>.git"
        }
    },
```
I don't see it even mention anything about tokens anymore, it just runs the git commands.

```
Executing command (CWD): git clone --mirror 'git@github.com:<vendor>/<repo>.git' '/home/<user>/.composer/cache/vcs/git-github.com-<vendor>-<repo>.git/'
Executing command (CWD): git clone --mirror 'https://***:***@github.com/<vendor>/<repo>.git' '/home/<user>/.composer/cache/vcs/git-github.com-<vendor>-<repo>.git/'
Executing command (/home/<user>/.composer/cache/vcs/git-github.com-<vendor>-<repo>.git/): git show-ref --tags --dereference
Executing command (/home/<user>/.composer/cache/vcs/git-github.com-<vendor>-<repo>.git/): git branch --no-color --no-abbrev -v
Executing command (/home/<user>/.composer/cache/vcs/git-github.com-<vendor>-<repo>.git/): git branch --no-color
Executing command (/home/<user>/.composer/cache/vcs/git-github.com-<vendor>-<repo>.git/): git show 'master':'composer.json'
Executing command (/home/<user>/.composer/cache/vcs/git-github.com-<vendor>-<repo>.git/): git -c log.showSignature=false log -1 --format=%at 'master'
```

**Comment by donquixote** [comment]
The `https://github.com/$vendor/$repo` -> `git@github.com:$vendor/$repo.git` can't be the full story.
The "no-api" + https did work at some time in the past, and it does work for my colleagues..

**Comment by Seldaek** [comment]
When using no-api, it does a clone attempt with the URL you pass it, so it really depends if your local env/git setup allows you to clone from https:// URLs or rather git@ URLs.

**Comment by stof** [comment]
could it be that `Username for 'https://github.com':` was asked by git rather than by composer itself ?

**Comment by donquixote** [comment]
@stof 
> could it be that Username for 'https://github.com': was asked by git rather than by composer itself ?

Yes, that would be a more accurate way to say it.
Of course it happens as part of `composer update --lock`.

So how would I test the oauth from the cli with just git, without composer?

**Comment by blackwolf12333** [comment]
This seems to be broken for me too.
